### DISCUSSION TITLE
feature request: edge device CUDA backend

### DISCUSSION THREAD

**Discussion by mindbeast** [pytorch/executorch#8214] [Ideas]
Does it make sense for executorch to have a mobile cuda backend? There are many edge devices in the Jetson lineup from nvidia that have a cuda gpu, but can benefit from not wanting to link an enormous libtorch dependence.

**Comment by bionictoucan** [comment]
+1 to this, would be nice to get comparable performance to TensorRT without having to export models to ONNX etc. first!

**Comment by hietalajulius** [comment]
+1

**Comment by mergennachin** [comment]
@mindbeast @bionictoucan @hietalajulius 

Hi, thanks for the comment. 

Yes, that makes sense in general. 

Right now, for ExecuTorch, we are integrating Vulkan into ExecuTorch. The reason is that it is a suitable solution for mobile GPUs. Enabling mobile use-cases is our primary goal at the moment.

We will revisit Cuda, but perhaps, in the second half in the year. Curious, what are your current product needs?

**Comment by DzAvril** [comment]
Apologies for opening a similar feature request in #5263.

> Curious, what are your current product needs?

@mergennachin We want to deploy LLMs in cars, but Python-based inference frameworks like vLLM and SGLang are not suitable for edge devices. 

> We will revisit Cuda, but perhaps, in the second half in the year. 

Nearly five months have passed, is there any progress on this?

**Comment by digantdesai** [comment]
Thank you for following up @DzAvril.

> We want to deploy LLMs in cars, but Python-based inference frameworks like vLLM and SGLang are not suitable for edge devices.

I guess this is using a platform similar to Jetson?

> Nearly five months have passed, is there any progress on this?

No update yet on CUDA backend for ET at the moment. We can get back to you here once we plan something.

**Comment by DzAvril** [comment]
> I guess this is using a platform similar to Jetson?

@digantdesai Yes, Jetson Orin for now, and possibly Thor in the future.

Looking forward to your update.

**Comment by DuinoDu** [comment]
For mobile cuda backend, does `torch_tensorrt` satisfy the requirement?

**Comment by mindbeast** [comment]
@DuinoDu My expectation is that compatibility is poor with torch_tensorrt. I expect a more compliant backend in executorch would help a lot of developers.

**Comment by yrik** [comment]
Any update on this? What is the best alternative so far to run CUDA on jetson? Directly Torch? Onnx?

**Comment by shoumikhin** [comment]
@Gasoonjia may like to update this discussion?

**Comment by larryliu0820** [comment]
We have a WIP cuda backend, backend by AOTInductor: https://docs.pytorch.org/docs/stable/torch.compiler_aot_inductor.html. We have enabled some popular models (whisper, voxtral, gemma3 etc), please checkout this [README.md](https://github.com/pytorch/executorch/blob/main/examples/models/whisper/README.md) to give whisper a try! You can also find voxtral instructions here: https://github.com/pytorch/executorch/tree/main/examples/models/voxtral#readme, gemma3 instructions here: https://github.com/pytorch/executorch/blob/main/examples/models/gemma3/README.md

**Reply by mindbeast** [reply]
You should actually link to the backend: https://github.com/pytorch/executorch/blob/60141298948e27ef5527351e0260a24fd2831bf2/backends/cuda/cuda_backend.py#L31

Why is executorch involved at all if you are just using the AOTInductor backend from torch? Any overview of design here?

I don't know what whisper models are and don't see how that is related to the CUDA backend.

**Reply by larryliu0820** [reply]
Feel free to use AOTInductor directly. 

> Any overview of design here

This work stream is still experimental/exploratory, so no concrete design just yet. We are evaluating several different options (Vulkan, TensorRT, ggml etc) alongside with AOTInductor. 

> I don't know what whisper models are 

Just some model to showcase we have the e2e working. You don't have to use it.

**Reply by phamtrongthang123** [reply]
I gave your whisper README a try and it works!

**Reply by WhoisZihan** [reply]
Looking forward to it, but how do you deal with the custom ops written in triton/c++/cuda？

**Reply by Gasoonjia** [reply]
@WhoisZihan you can check for our current way for triton (https://github.com/pytorch/executorch/tree/main/backends/cuda/triton) and cuda (https://github.com/pytorch/executorch/blob/main/backends/cuda/runtime/shims/int4mm.cu). 
For c++ do you mean you wanna deploy part of models in cpu?

**Reply by WhoisZihan** [reply]
> @WhoisZihan you can check for our current way for triton (https://github.com/pytorch/executorch/tree/main/backends/cuda/triton) and cuda (https://github.com/pytorch/executorch/blob/main/backends/cuda/runtime/shims/int4mm.cu). For c++ do you mean you wanna deploy part of models in cpu?

@Gasoonjia Thanks for the reference, for c++ I mean I have a model with some custom triton/cuda/c++ ops, and would like to export & compile it, then deploy the compiled object for inference in C++, either in executorch or libtorch.

The native torch AOTI compilation seems not including the custom ops by default, therefore I cannot reload it in libtorch.

I wonder if executorch compiles these custom ops written in c++/cuda, so that I can distribute the compile object to another machine and reload it? Or do I have to implement the same custom op in executorch/libtorch in order to run inference?

**Reply by Gasoonjia** [reply]
Sry @WhoisZihan i missed your reply.
I suppose you want to run your model using your custom op and triton op? not sure if i understand your target correctly. 
For triton op, you can follow our current way to register your triton op (https://github.com/pytorch/executorch/blob/main/backends/cuda/triton/replacement_pass.py#L25)
For register other custom op, you can also reuse the AOTI api (e.g. https://github.com/pytorch/pytorch/pull/164212)

**Reply by Gasoonjia** [reply]
For the compile code, they are in c++/cuda and will be compiled into a .so-like file. You can directly run it on target device without reimpl it.
